### PR TITLE
ECC-1118| migration to hmrc-mongo, updating dependencies WIP

### PR DESCRIPTION
**Pull request by oliwiaKot**

### CODE REVIEW

**Review by darbyo — CHANGES_REQUESTED** [review]

File: app/uk/gov/hmrc/eoricommoncomponent/frontend/services/cache/sessionCache.scala [hunk 1]
```diff
@@ -45,50 +44,10 @@ sealed case class CachedData(
...
-  private def throwException(name: String, sessionId: Id) =
-    throw new IllegalStateException(s"$name is not cached in data for the sessionId: ${sessionId.id}")
-
-  private val initialEmptySubscriptionDetails = SubscriptionDetails()
-}
+)
 
 object CachedData {
-  val regDetailsKey                    = "regDetails"
+  val regDetailsKey                    = "regDetails"     //
```
[the remark this answers is not in the document]
Comment (oliwiaKot) [reply app/uk/gov/hmrc/eoricommoncomponent/frontend/services/cache/sessionCache.scala]: typo, removed now

File: app/uk/gov/hmrc/eoricommoncomponent/frontend/services/cache/sessionCache.scala [hunk 2]
```diff
@@ -98,128 +57,157 @@ object CachedData {
   val keepAliveKey                     = "keepAlive"
   val safeIdKey                        = "safeId"
   val groupIdKey                       = "cachedGroupId"
-  val groupEnrolmentKey                = "groupEnrolment"
+  val groupEnrolmentKey                = "groupEnrolment" //
```
[the remark this answers is not in the document]
Comment (oliwiaKot) [reply app/uk/gov/hmrc/eoricommoncomponent/frontend/services/cache/sessionCache.scala]: removed

File: app/uk/gov/hmrc/eoricommoncomponent/frontend/controllers/GYEHowCanWeIdentifyYouNinoController.scala [hunk 3]
```diff
@@ -96,8 +99,8 @@ class GYEHowCanWeIdentifyYouNinoController @Inject() (
     )
   }
 
-  // TODO Get rid of `.get`. Now if there is no information Exception will be thrown, understand what should happen if this is not provided
-  private def retrieveNameDobFromCache()(implicit hc: HeaderCarrier): Future[Individual] =
+  // TODO Get rid of `.get`. Now if there is no information Exception will be thrown, understand what must happen if this is not provided
```
[the remark this answers is not in the document]
Comment (oliwiaKot) [reply app/uk/gov/hmrc/eoricommoncomponent/frontend/controllers/GYEHowCanWeIdentifyYouNinoController.scala:102]: ticket created https://jira.tools.tax.service.gov.uk/browse/ECC-1283

**Comment by beckfordp** [comment]
Once we made the decision to replace the SessionCache with on that extends SessionCacheRepository from hmrc-mongo then the interface to the cache changes and this change becomes huge. I questioned the value of doing so at the time when we did it in subscription... but its done now, and at least our session caches are now consistent with each other and the rest of hmrc :)

- checked that the hmrc-mongo SessionCache was being used

**Comment by beckfordp** [comment]
Once we made the decision to change the interface to the SessionCache by extending SessionCacheRepository from hmrc-mongo, this refactor became huge :) I questioned the value at the time when we did so in subscription... but its done now, and both our session caches are now consistent with each other and other projects that use hmrc-mongo.

- Checked that the SessionCache now extends SessionCacheRepository from hmrc-mongo
- Reviewed the ripple effect the change in the SessionCache interface has caused (implicit request params etc).
- All compiles, and the tests are all passing locally.

LGTM 👍

File: app/uk/gov/hmrc/eoricommoncomponent/frontend/services/cache/sessionCache.scala [hunk 4]
```diff
@@ -106,120 +65,153 @@ object CachedData {
...
 
-  def eori(implicit hc: HeaderCarrier): Future[Option[String]] =
-    getCached[Option[String]](sessionId, (cachedData, _) => cachedData.eori)
+  def fetchSafeIdFromReg06Response(implicit request: Request[_]): Future[Option[SafeId]] =
+    registerWithEoriAndIdResponse.map(
+      response =>
+        response.responseDetail.flatMap(_.responseData.map(_.SAFEID))
+          .map(SafeId(_))
+    ).recoverWith {
+      case _ => Future.successful(None)
```
Comment (poscala) [inline app/uk/gov/hmrc/eoricommoncomponent/frontend/services/cache/sessionCache.scala:175]: I would recommend changing this to filter out fatal JVM exceptions as we shouldn't catch those, e.g. we need to let the app crash `OutOfMemmoryError`.

Can be achieved using `case NonFatal(e) => ...`

More info: https://stackoverflow.com/a/66118268, https://stackoverflow.com/questions/29744462/the-difference-between-nonfatal-and-exception-in-scala

File: app/uk/gov/hmrc/eoricommoncomponent/frontend/services/cache/sessionCache.scala [hunk 5]
```diff
@@ -106,120 +65,153 @@ object CachedData {
...
 
-  def groupEnrolment(implicit hc: HeaderCarrier): Future[EnrolmentResponse] =
-    getCached[EnrolmentResponse](sessionId, (cachedData, id) => cachedData.groupEnrolment(id))
+  def groupEnrolment(implicit request: Request[_]): Future[EnrolmentResponse] =
+    getData[EnrolmentResponse](groupEnrolmentKey).map(_.getOrElse(throwException(groupEnrolmentKey)))
+
+  def remove(implicit request: Request[_]): Future[Boolean] =
+    preservingMdc {
+      cacheRepo.deleteEntity(request).map(_ => true).recoverWith {
+        case _ => Future.successful(false)
```
Comment (poscala) [inline app/uk/gov/hmrc/eoricommoncomponent/frontend/services/cache/sessionCache.scala:207]: Same as in my other comment about handling Non Fatal JVM exceptions

**Review by poscala — APPROVED** [review]

**Review by beckfordp — APPROVED** [review]

**Review by darbyo — APPROVED** [review]

**Pull request merged**